### PR TITLE
LH consent

### PR DESCRIPTION
**Pull request by fisx**
https://wearezeta.atlassian.net/browse/SQSERVICES-402

see CHANGELOG.md for the jist of it.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,12 @@
 
 ## Release Notes
 
+   for on-prem folks who already use LH:
+    - change galley.yaml to FeatureLegalHoldEnabledForWhitelist your teams
+    - that's it!  no other changes needed.  (i think?)
```
Comment (jschaul) [inline CHANGELOG.md:9]: Could you reword these sentences to be less informal and not use abbreviations? The `FeatureLegalHoldEnabledForWhitelist` should be documented how to use; and release notes should link to that documentation.

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 2]
```diff
@@ -258,11 +259,13 @@ instance ToJSON FeatureSSO where
 instance FromJSON FeatureLegalHold where
   parseJSON (String "disabled-permanently") = pure $ FeatureLegalHoldDisabledPermanently
   parseJSON (String "disabled-by-default") = pure $ FeatureLegalHoldDisabledByDefault
+  parseJSON (String "whiteliste-teams-and-implicit-consent") = pure FeatureLegalHoldWhitelistTeamsAndImplicitConsent
```
Comment (jschaul) [inline libs/galley-types/src/Galley/Types/Teams.hs:262]:
```suggestion
  parseJSON (String "whitelisted-teams-and-implicit-consent") = pure FeatureLegalHoldWhitelistedTeamsAndImplicitConsent
```

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 3]
```diff
@@ -258,11 +259,13 @@ instance ToJSON FeatureSSO where
...
   parseJSON bad = fail $ "FeatureLegalHold: " <> cs (encode bad)
 
 instance ToJSON FeatureLegalHold where
   toJSON FeatureLegalHoldDisabledPermanently = String "disabled-permanently"
   toJSON FeatureLegalHoldDisabledByDefault = String "disabled-by-default"
+  toJSON FeatureLegalHoldWhitelistTeamsAndImplicitConsent = String "whiteliste-teams-and-implicit-consent"
```
Comment (jschaul) [inline libs/galley-types/src/Galley/Types/Teams.hs:268]:
```suggestion
  toJSON FeatureLegalHoldWhitelistedTeamsAndImplicitConsent = String "whitelisted-teams-and-implicit-consent"
```

File: services/galley/src/Galley/API/LegalHold.hs [hunk 4]
```diff
@@ -56,16 +58,23 @@ import UnliftIO.Async (pooledMapConcurrentlyN_)
...
+      pure False
+    FeatureLegalHoldDisabledByDefault -> do
+      statusValue <- Public.tfwoStatus <$$> TeamFeatures.getFeatureStatusNoConfig @'Public.TeamFeatureLegalHold tid
+      return $ case statusValue of
+        Just Public.TeamFeatureEnabled -> True
+        Just Public.TeamFeatureDisabled -> False
+        Nothing -> False
+    FeatureLegalHoldWhitelistTeamsAndImplicitConsent -> do
+      view (options . Opts.optSettings . Opts.setLegalHoldTeamsWhitelist)
+        <&> maybe (False {- reasonable default, even though this is impossible -}) (tid `elem`)
```
Comment (jschaul) [inline services/galley/src/Galley/API/LegalHold.hs:77]: Maybe: impossible due to option validation happening at server startup in galley.Options

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 5]
```diff
@@ -594,24 +650,151 @@ testGetTeamMembersIncludesLHStatus = do
...
+      liftIO $ assertEqual "" lhs' UserLegalHoldDisabled
+
+    do
+      -- nobody can request LH device.
+      withLHWhitelist tid' (requestLegalHoldDevice' g owner member tid) !!! testResponse 403 (Just "legalhold-not-enabled")
+      UserLegalHoldStatusResponse userStatus _ _ <- withLHWhitelist tid' (getUserStatusTyped' g member tid)
+      liftIO $
+        assertEqual
+          "requestLegalHoldDevice should set user status to Pending"
+          UserLegalHoldDisabled
```
Comment (jschaul) [inline services/galley/test/integration/API/Teams/LegalHold.hs:721]: The string in assertEqual doesn't seem to reflect what is being tested here. Equality is tested for Disabled, not Pending. Am I holding this wrong or could this be made clearer / the text changed?

**Review by jschaul — COMMENTED** [review]
Implementation looks reasonable with respect to the jira issue.

It would be nice to either add documentation, or link to an issue explaining the motivation to A) set up an endpoint for granting consent, but B) never using it but instead assuming implicit consent for all the users in the whitelist.

Could you document what is meant by consent exactly? consent to have a team admin enable a LH device for oneself? Or consent to exchange messages with someone who has a LH device?

See nitpicks inline.

File: libs/galley-types/src/Galley/Types/Teams.hs (hunk 2 above)
Comment (fisx) [reply libs/galley-types/src/Galley/Types/Teams.hs:262]: Thanks for the typo, but I prefer imperative over passive (am I getting the terminology right here?)

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 5 above)
Comment (fisx) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: copy&paste error, thanks!  35f487e80563706f5a60532161683a671405213c

File: services/galley/src/Galley/API/LegalHold.hs (hunk 4 above)
Comment (fisx) [reply services/galley/src/Galley/API/LegalHold.hs]: 8bcd0936d0975cfc68b9b9e5804c0692e18c4ad5

File: libs/galley-types/src/Galley/Types/Teams.hs (hunk 2 above)
Comment (fisx) [reply libs/galley-types/src/Galley/Types/Teams.hs]: 0566e30500d6d1d3c5645253d4aa49649e61215f

File: CHANGELOG.md (hunk 1 above)
Comment (fisx) [reply CHANGELOG.md]: 9af7b37829de11becf25e78203de16e30ac90a55

(Obviously this wasn't ready, i also mentioned yesterday to you that I am going to elaborate on that, but always to go have a reminder.)

**Comment by fisx** [comment]
> It would be nice to either add documentation, or link to an issue explaining the motivation to A) set up an endpoint for granting consent, but B) never using it but instead assuming implicit consent for all the users in the whitelist.
> 
> Could you document what is meant by consent exactly? consent to have a team admin enable a LH device for oneself? Or consent to exchange messages with someone who has a LH device?

good questions, thanks!  f34c52a7e9fde148cdaaf64dac5e24d56ef0cc5a

**Comment by fisx** [comment]
### Rationale: why put the whitelisted teams into `galley.yaml`, and not in a table in cassandra?

- I thought it was slightly faster, involving changing slightly fewer parts.
- It's safer in the sense that accidental whitelisting is harder (it requires more eyes).
- It makes it more explicit that it is a work-around to provide a feature to certain teams that we ultimately want to roll out to everybody.  Once we get there, we can remove the whitelists entirely.
- Performance of the whitelisting operation is less of an issue: if we decide to block / kick users based on the fact that LH allowed (without any LH devices being enabled yet), we can do that when galley is deployed, and before it comes fully online.  This will avoid the risk of race conditions and high load of production servers.

**Review by jschaul — APPROVED** [review]

**Pull request merged**